### PR TITLE
bootstrap non-initializing shard first

### PR DESCRIPTION
**Pull request by jshencode**
1. bootstrap non initializing (available and leaving) shards first
2. on server start, only wait for non initializing shards readiness

### CODE REVIEW

File: datanode/datanode.go [hunk 1]
```diff
@@ -302,83 +306,140 @@ func (d *dataNode) startActiveTopologyWatch() {
 	}
 }
 
-func (d *dataNode) startAnalyzingShardAvailability() {
-	ticker := time.NewTicker(5 * time.Second)
+func (d *dataNode) checkShardReadiness(tables []string, shards []uint32) (numReadyShards int) {
+	// check all whether all tables within initialing shards are bootstrapped
```
Comment (voyager-dw) [inline datanode/datanode.go]: update comment?

File: datanode/datanode.go [hunk 2]
```diff
@@ -302,83 +306,140 @@ func (d *dataNode) startActiveTopologyWatch() {
...
+				continue
+			}
+			if tableShard.IsBootstrapped() {
+				numTablesBootstrapped++
+			}
+			tableShard.Users.Done()
+		}
+
+		if numTablesBootstrapped == len(tables) {
+			shards[i], shards[numReadyShards] = shards[numReadyShards], shards[i]
```
Comment (voyager-dw) [inline datanode/datanode.go]: what purpose of this line?

File: datanode/datanode.go [hunk 3]
```diff
@@ -396,6 +457,9 @@ func (d *dataNode) ID() string {
 }
 
 func (d *dataNode) Serve() {
+	// wait for server is ready to serve
```
Comment (voyager-dw) [inline datanode/datanode.go]: add back the timer to see how much time take to recover?

File: datanode/datanode.go (hunk 2 above)
Comment (jshencode) [reply datanode/datanode.go]: swap two element

this is part of the partition algorithm, which partitions the list into two parts, ready shards and non ready shards

File: datanode/datanode.go (hunk 3 above)
Comment (jshencode) [reply datanode/datanode.go]: it is already in the Serve function

**Pull request merged**